design/protocol-changes/vote-cert-diffusion.tex

```diff
@@ -1,79 +1,68 @@
 
+We now describe the dynamics of an honest node engaging as the client in vote diffusion with its peers, assuming that it is caught-up.
 \begin{itemize}
 \item
-  The size of a vote is constant so we do not need to transmit the size at any point, the client will be able to keep this in mind.
-
+  The node continually requests more vote IDs from its peers (with a limit on unacknowledged votes), using non-blocking or blocking requests depending on whether there are are outstanding unacknowledged votes.
 \item
-  We do not need to download the votes on the historical chain, so we can just start the protocol with the state of the server at that point, and there is no need to transmit an initial payload to explain since when we want to download votes.
+  At the beginning of a Peras round, the node will start receiving lots of new vote IDs for that round from all of its (honest) peers.
+  For each such vote ID, the node will ask one peer (or a small number in parallel) out of the peers that offered this vote ID for the corresponding vote, relying on protocol-level timeouts for a prompt delivery (or otherwise disconnecting from the peer).
+
+  By limiting the number of votes that are being downloaded in parallel, the traffic is implicitly spread out over the first few seconds of the round, bounding the spike of network activity.
 
+  This strategy naturally handles adversarial vote equivocation, cf.~\cref{sec:attack equivocations}.
 \item
-  We want to avoid missing a quorum when almost caught-up and when we put something in our immutable DB.\niols{I can't remember what I meant by that, but probably there is an important question of when we should start downloading votes. Is there a risk, if we want until we are caught-up?}
+  The client will \emph{not} download the vote corresponding to a vote ID if its round either lies in the distant past (\cref{sec:votes certs from the past}), or in the near past for which the node has already observed a successful quorum.
+\item
+  It might seem appealing to also stop downloading votes for the current round as soon as we have observed a quorum (via votes of weight \perasQuorum{}).
 
+  However, then the node can not relay these votes, and a downstream peer might be missing exactly some of these votes to observe a quorum of votes by itself.
+  Our design still handles this case directly by also running certificate diffusion between caught-up nodes (\cref{sec:certificate-diffusion}), but downloading and validating only a few votes is more efficient than transferring a certificate.\footnote{
+  It is conceivable that in a variant of~\cite{peras-cert-report}, certificates will be only slightly larger than votes, in which case the difference in efficiency might be negligible.}
+A compromise would be to stop downloading votes for the current round as soon as votes of weight larger than $\sigma$ have been observed, where $\perasQuorum < \sigma < 1$.
```

Comment:
I think I was thrown off by the use of the word “downloading”. Am I understanding correctly that the story here is:

> as a server for other peers, we might want to stop sending them votes as soon as we have a certificate, but let's not because they might prefer downloading our votes?

The word “download” makes me think of client behaviour, namely:

> as a client of other peers, I might want to stop requesting votes (bodies, not ids) as soon as I have a quorum

and that I think is reasonable?

Reply:
I was seeing ghosts here, probably confusing sth with equivocations. This section can be largely simplified.

The more general point about it being potentially desirable for downloading a few remaining votes instead of an entire certificates still makes sense.